BUG: Disallow type-promotion of dtypes with different field names

ahaldane opened the pull request:
Fixes #15494


The fix seems fairly painless. 



I also considered if we should restrict things even more, but didn't do so. For instance, what about if the field names and order are the same, but the byte-offsets of the fields are different. Which dtype "wins"?

Eg:

```python
>>> A = ("a", "<i8") 
>>> B = ("b", "<i8") 
>>> ab = np.rec.array(obj=np.array([]), dtype=[A, B]) 
>>> ba = np.rec.array(obj=np.array([]), dtype=[B, A])                                       
>>> np.concatenate([a_b_, b_a_[['a', 'b']]])                                                  
array([],
      dtype=(numpy.record, {'names':['a','b'], 'formats':['<i8','<i8'], 'offsets':[8,0], 'itemsize':16}))
```
The behavior shown here is actually the same as the behavior from 1.13 and before (with assign-by-fieldname), where the last dtype would be used in such cases, just like here. Since that is old behavior, I opted to leave it alone.

ahaldane committed:
BUG: Disallow type-promotion of dtypes with diff field names

Fixes #15494 #13667

eric-wieser commented:
Why does `ab[:] = b_a_` still work after this change?

ahaldane commented:
As I recall, casting (eg, `.astype`) is different from Type promotion (eg, ufunc binop rules), but I need to refresh myself a bit... 

This PR has doesn't affect `.astype`.

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -1357,6 +1357,17 @@ def test_multiindex_titles(self):
+        # gh-15494
+        A = ("a", "<i8")
+        B = ("b", "<i8")
+        ab = np.rec.array(obj=np.array([(1, 2), dtype='<i8']), dtype=[A, B])
```
Let's not throw `recarray` in here, the issue is more fundamental:
```suggestion
        ab = np.array([(1, 2)], dtype=[A, B])
```

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -1357,6 +1357,17 @@ def test_multiindex_titles(self):
+        A = ("a", "<i8")
+        B = ("b", "<i8")
+        ab = np.rec.array(obj=np.array([(1, 2), dtype='<i8']), dtype=[A, B])
+        ba = np.rec.array(obj=np.array([(1, 2), dtype='<i8']), dtype=[B, A])
```
```suggestion
        ba = np.array([(1, 2)], dtype=[B, A])
```

ahaldane commented:
Hmm, this was also touched in #13648. I haven't gone through it yet.

ahaldane commented:
Oh, yeah, this PR isn't quite right. It should totally ignore the field names, since can_cast should only depend on the field order.... working on it

eric-wieser commented:
> Oh, yeah, this PR isn't quite right. It should totally ignore the field names

Ah, but if we do that we'd be declaring the concatenate behavior as intentional, which I'm not sure is right - or at least, it wouldn't make the problem go away.

eric-wieser commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -629,35 +629,43 @@ type_num_unsigned_to_signed(int type_num)
+
+    val = PyObject_RichCompareBool(from->names, to->names, Py_EQ);
+    if (val != 1 || PyErr_Occurred()) {
+        PyErr_Clear();
```
I don;t think you should clear the error here, else `MemoryError` translates to "can't cast" which is nonsense.

eric-wieser commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -629,35 +629,43 @@ type_num_unsigned_to_signed(int type_num)
-    while (PyDict_Next(field1, &ppos, &key, &tuple1)) {
-        if ((tuple2 = PyDict_GetItem(field2, key)) == NULL) {
+    while (PyDict_Next(from->fields, &ppos, &key, &tuple1)) {
+        if ((tuple2 = PyDict_GetItem(to->fields, key)) == NULL) {
```
```suggestion
        if ((tuple2 = PyDict_GetItemWithError(to->fields, key)) == NULL) {
            if (PyErr_Occurred()) {
                return -1;
            }
```

ahaldane commented:
It seems like we want for two dtypes with different field names but same field dtypes:
 * can be cast to each other
 * cannot be promoted to a common type unambiguously

So can_cast should succeed, but promote_types should fail. Does that make sense?

ahaldane commented:
All right. So I fixed up `can_cast_fields` to  ignore the field names when determining if a structured cast is possible, which is more consistent with the "*-by-position" strategy we use now.

There was a complication, that when doing `a == b` on two structured arrays, numpy currently has some half-deprecated code which tested whether a and b were castable (rather than promotable) before deciding to try to do elementwise comparison or to fail and return a scalar. Because more things now count as castable, we fall into the elementwise case more often, and it turns out the elementwise case had some special-case code for structured arrays which was still using the old *-by-fieldname pattern in `_void_compare`. So I fixed up `_void_compare` to do comparisons by-field-position.

I also added a paragraph to the structured array docs about all this.

ahaldane commented on numpy/testing/tests/test_utils.py:
```diff
@@ -152,7 +152,8 @@ def test_recarrays(self):
+        c = np.empty(2, [('floupipi', float),
+                         ('floupi', float), ('floupa', float)])
         c['floupipi'] = a['floupi'].copy()
         c['floupa'] = a['floupa'].copy()
```
This change was needed because now the comparison 3 lines below here falls into the elementwise comparison case, since b and c were now castable after thie PR. So here I added an extra field to make sure b and c are not castable so they trip the warning.

eric-wieser commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -1088,6 +1099,16 @@ PyArray_PromoteTypes(PyArray_Descr *type1, PyArray_Descr *type2)
+            return NULL;
+        }
+        Py_INCREF(type1);
+        return type1;
```
Can you add a test for this behavior? In that if two dtypes with different padding are specified, the first/last one is returned by identity?

ahaldane replied:
Actually the way I wrote it using "equivalent_fields"  this rejects promotion for dtypes with different padding/offsets. Probably that's safer anyway so no-one gets their padding changed unexpectedly.

Added a test for that case.

eric-wieser commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -629,44 +629,55 @@ type_num_unsigned_to_signed(int type_num)
+        PyArray_Descr *to_fld_dtype, *from_fld_dtype;
+
+        to_key = PyTuple_GET_ITEM(to->names, i);
+        if ((to_tup = PyDict_GetItemWithError(to->fields, to_key)) == NULL) {
```
```suggestion
        PyObject *to_tup = PyDict_GetItemWithError(to->fields, to_key))
        if (to_tup == NULL) {
```
and similar below

eric-wieser reviewed (commented):
Generally looks good, thanks!

Some style comments, and a test I think we missed.

eric-wieser commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -629,44 +629,55 @@ type_num_unsigned_to_signed(int type_num)
+            if (!PyErr_Occurred()) {
+                PyErr_SetString(PyExc_ValueError, "bug, should not happen");
+            }
+            return 0;
```
Hmm, guess the caller doesn't support this

eric-wieser replied:
In which case, you'll need to call `PyErr_Clear()` here instead.

I've a local branch that tries to propagate errors out of here, but it's a bit invasive, and less important than this patch.

ahaldane commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -629,44 +629,55 @@ type_num_unsigned_to_signed(int type_num)
-                        (PyArray_Descr *)PyTuple_GET_ITEM(tuple1, 0),
-                        (PyArray_Descr *)PyTuple_GET_ITEM(tuple2, 0),
-                        casting)) {
+        if (!PyArray_CanCastTypeTo(from_fld_dtype, to_fld_dtype, casting)) {
```
I left this one, since we shouldn't short circuit here

eric-wieser replied:
Whoops, good point.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1114,29 +1114,43 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
+        field_count = PyTuple_GET_SIZE(self_descr->names);
+        if (field_count != PyTuple_GET_SIZE(other_descr->names)) {
+            PyErr_SetString(PyExc_RuntimeError, "bug, should never happen.");
+            return 0;
```
```suggestion
            return NULL;
```

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1114,29 +1114,43 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
-        Py_ssize_t pos = 0;
+        PyArray_Descr *self_descr, *other_descr;
+        int i, field_count;
+
```
nit: Mind moving some or all of these down to where they are initialized?

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -1357,6 +1357,34 @@ def test_multiindex_titles(self):
+        B = ("b", "<i8")
+        ab = np.array([(1, 2)], dtype=[A, B])
+        ba = np.array([(1, 2)], dtype=[B, A])
+        assert_raises(TypeError, np.concatenate, [ab, ba])
```
What's the behavior of the following?
```suggestion
        assert_raises(TypeError, np.concatenate, [ab, ba])
        assert_raises(TypeError, np.concatenate, [ab[['a', 'b']], ba[['a', 'b']]])
```

eric-wieser replied:
Ah, I guess that's covered below

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -1357,6 +1357,34 @@ def test_multiindex_titles(self):
+        # dtypes with same field names/order but different memory offsets
+        # are not promotable either.
+        assert_raises(TypeError, np.promote_types,
+                                 [ab.dtype, ba[['a', 'b']].dtype])
```
I imagine this is going to be a bit annoying for when I try to do something like `np.concatenate((arr1[['a', 'b']], arr2[['a', 'b']]))`.

A reasonable alternative might be for `promote_types` to eliminate all padding if it doesn't match.

But maybe we can deal with that in a follow-up.

ahaldane replied:
That seems reasonable. I think a counterargument might be that we want to avoid adding special behaviors/cases the user has to know about, and should instead to force the user to be specific, eg by doing
```python
np.concatenate((repack_fields(arr1[['a', 'b']]), repack_fields(arr2[['a', 'b']])))
```

Better yet, I think users should use something like a ducktyped-ArrayCollection instead of a structured array where padding is important.

In any case, let's disallow it in this PR.

seberg replied:
Rethinking, I actually would argue that promote types *should* eliminate all paddings. Since it normally returns a *new* dtype that can represent the same information as both input dtypes (safely cast both of them to this one).
There is some complexity about changing the padding when both input dtypes are identical (i.e. we currently ensure native byte order after promotion, but is a bit dubious).

In the future, I am thinking about adding a `to_canonical()` method, or so, which I suppose for structured dtypes could strip padding.

I have to think about what this actually changes API wise, I expect I think this fixes cases which are probably bugs only?

ahaldane replied:
The `promote_types` docstring does say "Returns the data type with the smallest size and smallest scalar kind ", which suggests that if you do `promote_types(a, a)` where a contains padding, it would remove the padding this would lead to the "smallest size".

So if we are dogmatically following the docstring, it does make sense to remove padding.

seberg replied:
I fully agree, but I would not give much heed to the docstring, it was obviously written with numerical types in mind :).

seberg commented:
@ahaldane how bad would it be to go through a deprecation here? And how buggy is the current behaviour? I do not mind putting this in, I expect it is rare enough that those running into it, can easily fix it. But, unless all cases that currently work are obvious bugs, we may want to not backport.

ahaldane commented:
I see three changes in this PR:
1. Type promotion: This PR raises an exception for type-promotion that worked before (backcompat breakage) but which would lead to weird results when the promoted type was used (#15494). That exception could easily and straightforwardly be replaced by warning, if we want a deprecation cycle. Many of the cases from the reports (eg the concatenate calls) affected by this raised FutureWarnings in 1.13 and the behaviors all changed in 1.14 already.
2. Casting: `np.can_cast` allows some casts that were disallowed before (no back-compat problem), but also disallows some (back-compat problem) involving re-ordered fields which can't be safely cast according to 'safe' rule. However, even in these cases, assignment and `.astype` have the same behavior as before, since they don't check `can_cast`. If we wanted a deprecation cycle, I think we would make `can_cast` raise a warning if the old method and the new method gave a different result.
3. Structure Comparison: because of the change in casting rules, a few cases which previously raised a warning and returned a scalar boolean will now return an elementwise boolean array.

While I expect problems will be quite unlikely, I really, really don't like having unhappy users report back. Let me try coding up a deprecation cycle commit on top of this.

seberg commented:
Lets not rush too much. I am trying to map out what actually changed `np.promote_types(np.dtype([("", "f4")]), np.dtype([("", "f8")]))` never worked. With this change it seems that:
```
np.can_cast(np.dtype([("a", "f4")]), np.dtype([("b", "f8")]))
```
I wonder if that should require `casting="unsafe"` (or `same_kind`), astype uses unsafe as well after all? But that is relaxing the behaviour.

It seems endianess changes in `promote_types` were passed through, but that seems like a pretty big corner case, so happy to deprecate, but if it gets ugly...

ahaldane commented:
For the endian changes, I just noticed that too, and that's easily fixed. If we are already changing endianness, that's further motivation for removing padding without asking.

ahaldane committed:
ENH: allow structure type-promotion if offsets differ

ahaldane commented:
I added a commit now where we promote to the smallest "packed" nbo datatype, instead of failing if the memory layout is different. It doesn't cause any test breakage here. Getting the endianness right took a try because by default, numpy treats structured datatypes as "|" (ignore) byte order so that `ensure_dtype_nbo` does nothing, so I ended up always calling `NewByteOrder(NATIVE)` in PromoteTypes.

I can take that commit out and leave it for another PR, but I put it here in case it helps think about what we want to do. Also, I saw that in #2676 a few users thumbed-up a request to allow type-promotion to ignore padding.

For your case `np.can_cast(np.dtype([("a", "f4")]), np.dtype([("b", "f8")]))`, that seems like a "safe" cast in the sense that we are not losing any precision and can always exactly cast back to the original dtype.

seberg commented:
Hmmm, for me "safe" should mean: No loss of information, so that I see renaming of fields as being at best a "same kind" operation. Of course that is debatable and "can roundtrip" is not unreasonable. My example would be `float64[meters] -> float64[unitless]` losing too much information to be considered "safe", which I feel is similar to the field renaming case (although that one I would define as "unsafe" for sure, while for field renaming I may think "same kind" is fine).

That definition does mean that convenient field renaming might have to rely on a utility function, somewhat outside the normal "casting".

eric-wieser commented:
I don't think that's the same meaning of "safe" that we currently have. To me, safe implies that `a.astype(b.dtype).astype(a.dtype)` loses no information.

What you're describing is loss of type information, not loss of value information. That's definitely a useful concept, but I'm not sure yet whether it deserves the same label.

seberg commented:
We can agree that safe implies `a.astype(b.dtype).astype(a.dtype)` loses no information. The question is whether it is a superset or a subset of "the value loses no information". We currently little notion of loss of type information, with the exception of `dt = np.dtype("timedelta64[s]"); np.array([3], dtype=dt).astype(np.int64).astype(dt)`. The best part of that: currently it is considered safe to cast from int64 to timedelta, but not in the reversed direction.

seberg commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -1374,6 +1384,47 @@ PyArray_PromoteTypes(PyArray_Descr *type1, PyArray_Descr *type2)
+            PyArray_Descr *ret = PyArray_DescrNewByteorder(type1, NPY_NATIVE);
+            Py_DECREF(type1);
+            return ret;
+        }
```
This is currently incorrect for:
```python
In [2]: dt1 = np.dtype([("", "i4")])                                                                                    
In [3]: dt2 = np.dtype([("", "i8")])                                                                                    
In [6]: np.promote_types(dt1, dt2)                                                                                      
Out[6]: dtype([('f0', '<i4')])

In [7]: np.promote_types(dt2, dt1)                                                                                      
Out[7]: dtype([('f0', '<i8')])
```
Which has to either not work for now, or probably promote each included dtype recursively.

ahaldane replied:
Yes, I totally missed this possibility.

seberg reviewed (commented):
I think I am OK with pushing this forward with the safe casting semantics as they are, the units are probably a somewhat different use-case. The more important thing is probably to see dtype promotion as not all that closely related to safe-casting semantics (even if we currently use that as a fallback in a few places).

Should we push this forward? I am unsure if it is important enough to push into a point release, or putting it into master only isn't the easier thing. The current type promotion logic needs to be fixed up though, either to be recursive, or simply more restrictive (for now).

ahaldane committed:
ENH: Allow type-promotion of structured dtypes field-by-field

seberg commented:
Just to note, I removed the 1.18 milestone, it is likely it will not exist, and even if, I am not certain we should backport this.

ahaldane commented:
Based on @sebergs last point, I added a new commit which allows type-promotion of structured dtypes so that corresponding fields are properly promoted, as in:

```python
>>> np.promote_types('i4,f8', 'i8,f4')                                      
dtype([('f0', '<i8'), ('f1', '<f8')])
```

Note that this is a new feature, such promotion was not allowed before.

I don't think this PR is worth a backport, we should just include it in the next release if we want it. You mentioned that if we don't want to backport, we may want to put another PR into 18.x to disallow the bad behavior, from #15494. I can look into that.

As to whether we want it, we should make sure all this new behavior is what really want before it is locked in. Personally, I think it seems like pretty reasonable behavior in the context of other structured array behavior, following the assign-by-field-position conventions we use now. The last commit only allows behavior that raised an exception before, so seems unlikely to cause downstream bugs. One weirdness with the new behavior is that concatenating an array to itself will strip any field-padding from the dtype, but that is similar to how np.concatenate strips all endianness info when you concatenate any array to itself.

seberg commented:
Lets not worry about backporting it seems more trouble then worth it. There seem to be two points we have to be clear on?

1. Promotion of structured dtypes will remove additional padding
2. We will use safe casting for casts from/to the unstructured equivalent datatype

right? For point 1. I think I am happy with it, the only reason to keep the padding is if you would see the padded areas as actual data that is merely hidden IMO. Point 2. is a bit less clear, but I am happy with it. I do see a chance that we will have things like named-structured-dtypes (i.e. you will register ufuncs for one). For such a DType I am not sure that I would allow safe casting. The main argument against it, is that I am not sure if safe-casting is actually useful. Normally users should (and have to) index the field to do anything useful in any case.

seberg commented:
@ahaldane 1.19 is coming around, should we try to finish this off, or prepare to put this into master once it is branched? For all I know this is pretty much good to go?

I am starting to grow on the idea that `"safe"` casting is maybe something we do not need to change. I.e. it being mainly a roundtripping property. The main advantage is that in that case, strings do not have to change their casting level. On the other hand, discoupling `promotion` and `safe-casting` seems good. (in theory someone could add an even `safer` casting level at some point I guess).
My arguments were always along parametric types (e.g. units) for which I do not like `safe` casting to work, but which are special enough to be different (strings may change their length, but the value roundtrips fine).

charris commented:
@ahaldane Ping.

charris commented:
Pushing off to 1.20.x.

charris commented:
Ping all, and rebase needed.

seberg commented:
This needs more than a rebase now :), there is an additional code path here on master which would also need updating.  Happy to review! But unless pinged explicitly, I will not prioritize this until the code that the current PR touches is *deleted* :).

ahaldane commented:
@seberg,  just to be clear, you recommend I wait before rebasing until you've completed more of the dtype updates?

I'm looking through the new master code now, I can see you have some TODOs in the code related to this PR.

seberg commented:
@ahaldane yeah, I specifically marked them. I am happy to rebase it now, it probably isn't much harder; We just have to change both the new and the old version (which probably just means keep the changes here and do the same changes in the new version).

mattip commented:
This should go in after #18676 (it will need some fixups)

seberg commented:
Closing, gh-19226 is the equivalent change taking into account that all of this has changed quite a lot.  (Unfortunately, there is currently still a small API addition missing for DTypes to finish the puzzle.)
